### PR TITLE
DB History Switcher, cleanup, mem leak fix

### PR DESCRIPTION
**Pull request by H3rnand3zzz**
# How to test the functionality
* `/help privacy`

# I ran valgrind when using my new feature
No

### CODE REVIEW

**Comment by H3rnand3zzz** [comment]
Don't mind failing CI, it's Github's problem (it couldn't fetch some files).

File: src/command/cmd_defs.c [hunk 1]
```diff
@@ -2717,6 +2717,21 @@ static const struct cmd_t command_defs[] = {
...
+
+    { CMD_PREAMBLE("/privacy",
+                   parse_args, 2, 3, NULL)
+      CMD_MAINFUNC(cmd_privacy)
+      CMD_TAGS(
+              CMD_TAG_CHAT)
+      CMD_SYN(
+              "/privacy logging on|redact|off")
+      CMD_DESC(
+              "Configure privacy settings. ")
```
Comment (jubalh) [inline src/command/cmd_defs.c:2729]: One space too much at the end.
And can you rephrase it to "Configure privacy related settings"?

File: src/command/cmd_defs.c [hunk 2]
```diff
@@ -2717,6 +2717,21 @@ static const struct cmd_t command_defs[] = {
...
+      CMD_ARGS(
+              { "logging on|redact|off", "Switch chat logging (including DB logging) [EXPERIMENTAL], might have unintended consequences when switched off, use with caution." })
```
Comment (jubalh) [inline src/command/cmd_defs.c:2731]: I would write "Switch chat logging. This will also disable logging in the internally used SQL database. Your messages will not be saved anywhere locally. This might have unintended consequences, like not being able to decrypt OMEMO encrypted messages received later via MAM, and should be used with caution."

**Comment by jubalh** [comment]
Made some comments but will review later in more detail.

**Review by jubalh — CHANGES_REQUESTED** [review]
See comments, plus commit 2db9457f13596cfa29b327fa734cd9c2124bc400 needs a reference to the issue.

File: src/command/cmd_defs.c (hunk 1 above)
Comment (H3rnand3zzz) [reply src/command/cmd_defs.c:2729]: > One space too much at the end. And can you rephrase it to "Configure privacy related settings"?

"Privacy Settings" is correct term, I am not sure why should it be renamed.

**Review by sjaeckel — CHANGES_REQUESTED** [review]
LGTM

Please squash that fixup commit into the `auto_gchar` one.

File: tests/unittests/test_preferences.c [hunk 3]
```diff
@@ -11,7 +11,7 @@
 void
 statuses_console_defaults_to_all(void** state)
 {
-    char* setting = prefs_get_string(PREF_STATUSES_CONSOLE);
+    gchar* setting = prefs_get_string(PREF_STATUSES_CONSOLE);
```
[the remark this answers is not in the document]
Comment (H3rnand3zzz) [reply tests/unittests/test_preferences.c:14]: It was somewhat problematic to auto_gchar in unittests. I don't remember exact reason, if simple include of common.h would suffice, I will change it.

[the remark this answers is not in the document]
Comment (sjaeckel) [reply tests/unittests/test_preferences.c:14]: Oh I overlooked that it's in the tests. NVM then

**Comment by H3rnand3zzz** [comment]
@jubalh review suggestions are applied, I hope that this PR can be merged soon so I can merge my big cleanup commit with it.

**Review by jubalh — APPROVED** [review]
I think all review comments were taken care of. Good to go

**Review by jubalh — CHANGES_REQUESTED** [review]
Actually https://github.com/profanity-im/profanity/pull/1857#pullrequestreview-1502731949 was not adressed.

The first commit still needs a reference to the issue about /privacy in the commit message.

**Review by jubalh — APPROVED** [review]

File: src/chatlog.c [hunk 4]
```diff
@@ -222,9 +222,13 @@ chat_log_msg_in(ProfMessage* message)
 }
 
 static void
-_chat_log_chat(const char* const login, const char* const other, const char* const msg,
+_chat_log_chat(const char* const login, const char* const other, const char* msg,
                chat_log_direction_t direction, GDateTime* timestamp, const char* const resourcepart)
 {
+    auto_gchar gchar* pref_dblog = prefs_get_string(PREF_DBLOG);
+    if (g_strcmp0(pref_dblog, "redact") == 0) {
```
Comment (sjaeckel) [inline src/chatlog.c:229]: I really don't like the fact that for each message we're going to look up the setting and do a `strcmp()` ... didn't we have that kind of discussion already?

File: src/database.c [hunk 5]
```diff
@@ -373,6 +374,14 @@ _get_message_enc_type(const char* const encstr)
 static void
 _add_to_db(ProfMessage* message, char* type, const Jid* const from_jid, const Jid* const to_jid)
 {
+    auto_gchar gchar* pref_dblog = prefs_get_string(PREF_DBLOG);
+
+    if (g_strcmp0(pref_dblog, "off") == 0) {
+        return;
+    } else if (g_strcmp0(pref_dblog, "redact") == 0) {
```
[the remark this answers is not in the document]
Comment (H3rnand3zzz) [reply src/database.c:381]: Why? https://github.com/profanity-im/profanity/blob/master/src/chatlog.c#L115 and many examples of this somewhat simple implementation in the preexisting code

File: src/database.c [hunk 6]
```diff
@@ -373,6 +374,14 @@ _get_message_enc_type(const char* const encstr)
...
+        message->plain = g_strdup("[REDACTED]");
```
[the remark this answers is not in the document]
Comment (jubalh) [reply src/database.c:382]: BTW @H3rnand3zzz  running valgrind on a new feature/pr when you try your new changes could show you such leaks before the review :)

[the remark this answers is not in the document]
Comment (H3rnand3zzz) [reply src/database.c]: > BTW @H3rnand3zzz running valgrind on a new feature/pr when you try your new changes could show you such leaks before the review :)

In my next PRs I already started to do it, but thanks for advise anyway.

**Review by jubalh — APPROVED** [review]

**Review by sjaeckel — APPROVED** [review]

**Comment by H3rnand3zzz** [comment]
Thanks for reviews! :)

**Comment by jubalh** [comment]
Thanks for your work as well
